packages/wdio-browserstack-service/src/testOps/listener.ts

```diff
@@ -0,0 +1,227 @@
+    DEFAULT_WAIT_TIMEOUT_FOR_PENDING_UPLOADS,
+    LOG_KIND_USAGE_MAP
+} from '../constants.js'
+import { sendScreenshots } from './requestUtils.js'
+import { BStackLogger } from '../bstackLogger.js'
+import * as util from 'node:util'
+import fs from 'node:fs'
+import path from 'node:path'
+
+class Listener {
+    private static instance: Listener
+    private readonly usageStats: UsageStats = UsageStats.getInstance()
+    private readonly testStartedStats: FeatureStats = this.usageStats.testStartedStats
+    private readonly testFinishedStats: FeatureStats = this.usageStats.testFinishedStats
+    private readonly hookStartedStats: FeatureStats = this.usageStats.hookStartedStats
+    private readonly hookFinishedStats: FeatureStats = this.usageStats.hookFinishedStats
+    private readonly cbtSessionStats: FeatureStats = this.usageStats.cbtSessionStats
+    private readonly logEvents: FeatureStats = this.usageStats.logStats
+    private requestBatcher?: RequestQueueHandler
+
+    private constructor() {
+    }
+
+    public static getInstance(): Listener {
+        if (!Listener.instance) {
+            Listener.instance = new Listener()
+        }
+        return Listener.instance
+    }
+
+    public async onWorkerEnd() {
+        await this.uploadPending()
+        await this.teardown()
+        this.saveData(this.usageStats.getDataToSave())
+    }
+
+    async uploadPending(waitTimeout = DEFAULT_WAIT_TIMEOUT_FOR_PENDING_UPLOADS, waitInterval = DEFAULT_WAIT_INTERVAL_FOR_PENDING_UPLOADS): Promise<unknown> {
+        // @ts-ignore
+        if (this.requestBatcher?.pendingUploads <= 0 || waitTimeout <= 0) { // TODO: remove ts-ignore
+            return
+        }
+
+        await sleep(waitInterval)
+        return this.uploadPending(waitTimeout - waitInterval)
+    }
+
+    async teardown() {
+        BStackLogger.debug('teardown started')
+        RequestQueueHandler.tearDownInvoked = true
+        await this.requestBatcher?.shutdown()
+        BStackLogger.debug('teardown ended')
+    }
+
+    public hookStarted(hookData: TestData): void {
+        try {
+            this.hookStartedStats.triggered()
+            this.sendBatchEvents(this.getEventForHook('HookRunStarted', hookData))
+        } catch (e) {
+            this.hookStartedStats.failed()
+            throw e
+        }
+    }
+
+    public hookFinished(hookData: TestData): void {
+        try {
+            this.hookFinishedStats.triggered(hookData.result)
+            this.sendBatchEvents(this.getEventForHook('HookRunFinished', hookData))
+        } catch (e) {
+            this.hookFinishedStats.failed(hookData.result)
+            throw e
+        }
+    }
+
+    public testStarted(testData: TestData): void {
+        try {
+            this.testStartedStats.triggered()
+            this.sendBatchEvents(this.getEventForHook('TestRunStarted', testData))
+        } catch (e) {
+            this.testStartedStats.failed()
+            throw e
```

Comment:
Are we catching this ?

Reply:
Yes the functions calling it will catch the errors